sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java

```diff
@@ -149,6 +151,13 @@ public void populateDisplayData(DisplayData.Builder builder) {
     assertThat(unboundedDisplayData, hasDisplayItem("maxReadTime", maxReadTime));
   }
 
+  @Test
+  public void testReadBoundedTypeDescriptor() {
+    PCollection<String> input = pipeline.apply(Read.from(new SerializableBoundedSource()));
+    TypeDescriptor<String> typeDescriptor = input.getTypeDescriptor();
+    assertEquals(String.class, typeDescriptor.getType());
```

Comment:
Do you want to try a test where the source has a generic variable?

Reply:
Just added the case with generics, please tell me if this is what you had in mind.